File: java/src/main/java/com/ngt/state/QueryableStateDemo.java
package com.ngt.state;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.RichMapFunction;
import org.apache.flink.api.common.restartstrategy.RestartStrategies;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.common.time.Time;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.configuration.QueryableStateOptions;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;
import org.apache.hadoop.yarn.webapp.hamlet2.Hamlet;

/**
 * @author ngt
 * @create 2021-02-01 14:52
 * https://ci.apache.org/projects/flink/flink-docs-release-1.12/zh/dev/stream/state/queryable_state.html
 */
public class QueryableStateDemo {
    public static void main(String[] args) throws Exception {
        Configuration conf = new Configuration();
        conf.setInteger("rest.port", 8181);
        conf.setBoolean(QueryableStateOptions.ENABLE_QUERYABLE_STATE_PROXY_SERVER, true);

        StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(conf);
        env.enableCheckpointing(5000);
        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(3, Time.seconds(3)));
        DataStreamSource<String> lines = env.socketTextStream("192.168.31.8", 8888);

        SingleOutputStreamOperator<Tuple2<String, Integer>> words = lines.flatMap(new FlatMapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public void flatMap(String s, Collector<Tuple2<String, Integer>> collector) throws Exception {
                String[] words = s.split(" ");
                for (String word : words) {
                    collector.collect(Tuple2.of(word, 1));
                }
            }
        });

        KeyedStream<Tuple2<String, Integer>, String> keyed = words.keyBy(new KeySelector<Tuple2<String, Integer>, String>() {
            @Override
            public String getKey(Tuple2<String, Integer> tp) throws Exception {
                return tp.f0;
            }
        });

        SingleOutputStreamOperator<Tuple2<String, Integer>> summed = keyed.sum(1);
        keyed.map(new MyQueryStateRichMapFunction()).print();
        env.execute();

    }

    private static class MyQueryStateRichMapFunction extends RichMapFunction<Tuple2<String, Integer>, Tuple2<String, Integer>> {
        private transient ValueState<Integer> countState;

        @Override
        public void open(Configuration parameters) throws Exception {
            ValueStateDescriptor<Integer> stateDescriptor = new ValueStateDescriptor<>("wc-state", Integer.class);
            // 启用状态查询
            stateDescriptor.setQueryable("my-query-name");
            countState = getRuntimeContext().getState(stateDescriptor);
        }

        @Override
        public Tuple2<String, Integer> map(Tuple2<String, Integer> input) throws Exception {
            Integer count = input.f1;
            // 获取当前Key对应的value的值
            Integer historyCount = countState.value();
            if (historyCount == null) {
                historyCount = 0;
            }
            int total = historyCount + count;
            // 更新状态中的数据
            countState.update(total);
            // 输出结果
            input.f1 = total;
            return input;
        }
    }
}

File: scala/src/main/scala/com/ngt/state/QueryableStateDemo.scala
package com.ngt.state

import org.apache.flink.api.common.functions.{FlatMapFunction, RichMapFunction}
import org.apache.flink.api.common.restartstrategy.RestartStrategies
import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.api.common.time.Time
import org.apache.flink.configuration.{Configuration, QueryableStateOptions}
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

/**
 * @author ngt
 * @create 2021-02-09 8:22
 */
object QueryableStateDemo {
  def main(args: Array[String]): Unit = {
    val conf: Configuration = new Configuration()
    conf.setInteger("rest.port", 8181)
    conf.setBoolean(QueryableStateOptions.ENABLE_QUERYABLE_STATE_PROXY_SERVER, true)

    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(conf)

    env.enableCheckpointing(5000);
    env.setRestartStrategy(RestartStrategies.fixedDelayRestart(3, Time.seconds(3)));
//    val lines: DataStream[String] = env.socketTextStream("192.168.31.8", 8888)
    val lines: DataStream[String] = env.socketTextStream("192.168.31.8", 8888)

    val wordAndOne: DataStream[(String, Int)] = lines.flatMap(new FlatMapFunction[String, (String, Int)] {
      override def flatMap(value: String, out: Collector[(String, Int)]): Unit = {
        val words: Array[String] = value.split(" ")
        for (elem <- words) {
          out.collect((elem, 1))
        }
      }
    })

    wordAndOne.keyBy(_._1)
      .map(new MyQueryStateRichMapFunction)

  }

  case class MyQueryStateRichMapFunction() extends RichMapFunction[(String, Int), (String, Int)] {

    private var countState: ValueState[Int] = _


    override def open(parameters: Configuration): Unit = {
      val stateDescriptor: ValueStateDescriptor[Int] =
        new ValueStateDescriptor[Int]("wc-state", classOf[Int])

      stateDescriptor.setQueryable("my-query-name")
      countState = getRuntimeContext().getState(stateDescriptor)
    }

    override def map(value: (String, Int)): (String, Int) = {
      val count: Int = value._2
      var historyCount: Int = countState.value()
      if (historyCount == null) {
        historyCount = 0
      }
      historyCount += count
      countState.update(historyCount)
      (value._1, historyCount)
    }
  }

}
